### PR TITLE
[CCXDEV-13027] Add sequence diagram for canary rollouts

### PR DESCRIPTION
**Pull request by juandspy**
# Description

Add sequence diagram for canary rollouts

## Type of change

- Documentation update

## Checklist
* [ ] `make before_commit` passes
* [ ] updated documentation wherever necessary
* [ ] added or modified tests if necessary
* [ ] updated schemas and validators in [insights-data-schemas](https://github.com/RedHatInsights/insights-data-schemas) in case of input/output change

### CODE REVIEW

File: docs/canary_releases.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        RT ->> AI: Revert the PR
+        AI ->> GCS: Update prod deployment
+        GCS -->> C: Get version 1.0.1
+        GCS -->> CC: Get version 1.0.1
+    else canary archives look fine
+        RT ->> AI: Remove the canary version and set stable <br/>CANARY_VERSION = none<br/>STABLE_VERSION = 1.0.2
+        AI ->> GCS: Update prod deployment
+        GCS -->> C: Get version 1.0.2
+        GCS -->> CC: Get version 1.0.2
+        RT ->> Unleash: Set the canary population to 0%
```
Comment (ikerreyes) [inline docs/canary_releases.md:57]: Is this compatible with `Remove the canary version and set stable <br/>CANARY_VERSION = none<br/>STABLE_VERSION = 1.0.2`  in an async mode?

File: docs/canary_releases.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+```mermaid
+sequenceDiagram
+    actor RT as Rules Team
+    actor PT as Processing Team
+    participant GC as Gathering Conditions
+    participant AI as App Interface
+    participant GCS as Gathering Conditions Service
+    participant Unleash
+    participant Storage
```
Comment (ikerreyes) [inline docs/canary_releases.md:16]: More than storage I guess it will be the monitoring thing

File: docs/canary_releases.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    RT ->> Unleash: Increase the canary population to X%
+    GCS -->> C: Get version 1.0.1
+    rect rgb(143,188,143)
+        GCS -->> CC: Get version 1.0.2
+    end
+    
+    RT ->> Storage: Analyze the canary archives and check they are generated well
+
+    alt canary archives are wrong
+        RT ->> AI: Revert the PR
```
Comment (ikerreyes) [inline docs/canary_releases.md:48]: If you revert the PR, the canary version will be `none`? Then what happens with the population label as "canary"?

**Review by ikerreyes — COMMENTED** [review]
Do we need to prevent that if a canary is in progress another steps in?

Another question, what will be the cost and risk of having unleash set to X% all the time and we have always 2 deployments and we just change the versions of those?

File: docs/canary_releases.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    
+    rect rgb(250,128,114)
+        alt canary or stable version not found
+            GCS ->> GCS: CrashLoopBackOff
+            PT ->> AI: Revert the PR
+        end
+    end
+
+    GCS -->> C: Get version 1.0.1
+    GCS -->> CC: Still get version 1.0.1
```
Comment (jholecek-rh) [inline docs/canary_releases.md:37]: This assumes the canary population setting was at 0%, but the "canary archives are wrong" doesn't reset it.

File: docs/canary_releases.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    actor PT as Processing Team
+    participant GC as Gathering Conditions
+    participant AI as App Interface
+    participant GCS as Gathering Conditions Service
+    participant Unleash
+    participant Storage
+    actor C as Cluster
+    actor CC as Canary Cluster
+
+    Note left of GCS: Latest version is 1.0.1
```
Comment (jholecek-rh) [inline docs/canary_releases.md:20]: Let's express this consistently in terms of the two variables to avoid ambiguity. I suppose this means that the deployment uses `STABLE_VERSION = 1.0.1` and `CANARY_VERSION` is not set or is empty. This would answer @ikerreyes's [question](https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/pull/234/files#r1675730125).

Comment (jholecek-rh) [reply docs/canary_releases.md:20]: We might also state the prerequisite here that the canary population in Unleash is supposed to be at 0% at this point.

File: docs/canary_releases.md (hunk 3 above)
Comment (jholecek-rh) [reply docs/canary_releases.md:48]: The happy day scenario below also has a period of time where `CANARY_VERSION = none` while some clusters still get the canary label. It suggests that the service serves the `STABLE_VERSION`.

File: docs/canary_releases.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    end
+```
+
+In case of an emergency release, we would just need to create the app-interface
+Pull Request and ignore the Unleash settings.
+
+Note that the stage environment is not mentioned in the whole diagram. The
+idea is that the service logic is tested both in stage and production, but the
+new conditions are shipped to production directly via environment variables. The
+service has to download the conditions on startup and fail if there is any issue,
```
Comment (jholecek-rh) [inline docs/canary_releases.md:67]: Could you elaborate on this, please? Does it mean the service would connect to GitHub and fetch the requested versions. I'm not sure I would like that. It creates a new attack surface and a new space for human mistakes (messing up tags) to propagate to production.

File: docs/canary_releases.md (hunk 3 above)
Comment (jholecek-rh) [inline docs/canary_releases.md:48]: I suppose PT would still have to review and merge the PR to revert the previous PR?

File: docs/canary_releases.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    
+    RT ->> Storage: Analyze the canary archives and check they are generated well
+
+    alt canary archives are wrong
+        RT ->> AI: Revert the PR
+        AI ->> GCS: Update prod deployment
+        GCS -->> C: Get version 1.0.1
+        GCS -->> CC: Get version 1.0.1
+    else canary archives look fine
+        RT ->> AI: Remove the canary version and set stable <br/>CANARY_VERSION = none<br/>STABLE_VERSION = 1.0.2
```
Comment (jholecek-rh) [inline docs/canary_releases.md:53]: I suppose this would be another PR. Could you describe this part at the same level of granularity as the first change, please?

**Comment by juandspy** [comment]
@ikerreyes 

> Do we need to prevent that if a canary is in progress another steps in?

As the canary is triggered by us, I don't see why there would be 2 canaries. Why would there be 2 in parallel? 

> Another question, what will be the cost and risk of having unleash set to X% all the time and we have always 2 deployments and we just change the versions of those?

I don't get this scenario: "we have always 2 deployments". Unleash is designed to be inside the code, so I'm not sure how to split traffic among 2 deployments. We would need to put a proxy in front of the gathering service and use unleash there to split the traffic among the canary and live versions. IMO, too much effort for the same benefits.

File: docs/canary_releases.md (hunk 1 above)
Comment (juandspy) [reply docs/canary_releases.md:57]: Yes. It can be done in parallel:

- Situation 1: you set the canary population to 0% -> the canary response won't be returned to any cluster.
- Situation 2: you keep the canary population to X% but remove the canary version: all the clusters, be it from the canary or live population, will consume the stable version

File: docs/canary_releases.md (hunk 2 above)
Comment (juandspy) [reply docs/canary_releases.md:16]: Yes. As it's yet to be decided I left the "Storage" naming. But it can be called differently, for sure, ty :)

File: docs/canary_releases.md (hunk 3 above)
Comment (juandspy) [reply docs/canary_releases.md:48]: I've already answered [here](https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/pull/234#discussion_r1677435200) but I'll copy here too:
> you keep the canary population to X% but remove the canary version: all the clusters, be it from the canary or live population, will consume the stable version

>The happy day scenario below also has a period of time where CANARY_VERSION = none while some clusters still get the canary label. It suggests that the service serves the STABLE_VERSION.

yes, exactly :)

File: docs/canary_releases.md (hunk 4 above)
Comment (juandspy) [reply docs/canary_releases.md:37]: No, even if the canary population is at X%, being X !=0, as the PR was reverted, all clusters, be it canary or normal, will receive the stable version (1.0.1).

File: docs/canary_releases.md (hunk 1 above)
Comment (ikerreyes) [reply docs/canary_releases.md:57]: OK, thanks. Can we have situation 2 described somewhere? To me it is not obvious that the system will behave like that

File: docs/canary_releases.md (hunk 5 above)
Comment (juandspy) [reply docs/canary_releases.md]: Updated

File: docs/canary_releases.md (hunk 3 above)
Comment (juandspy) [reply docs/canary_releases.md:49]: As it's a revert, it can be self-serviceable from the RT. I added the PT here too just in case

File: docs/canary_releases.md (hunk 7 above)
Comment (juandspy) [reply docs/canary_releases.md:55]: I added `PT ->> AI: Review and merge the PR`. Not sure what else to say. There are no more steps here. Once it's merged, it should be automatically deployed to prod.

**Comment by ikerreyes** [comment]
> @ikerreyes
> 
> > Do we need to prevent that if a canary is in progress another steps in?
> 
> As the canary is triggered by us, I don't see why there would be 2 canaries. Why would there be 2 in parallel?
> 
Because we might not control everything everyone is doing. If the process gets automated one day (hopefully), it can happen. Or even if manual, people doing the trigger and approval can be different than those who did for the 1st and the 2nd, or just by mistake. I want to know what will happen and what we can do to prevent, if needed.

> > Another question, what will be the cost and risk of having unleash set to X% all the time and we have always 2 deployments and we just change the versions of those?
> 
> I don't get this scenario: "we have always 2 deployments". Unleash is designed to be inside the code, so I'm not sure how to split traffic among 2 deployments. We would need to put a proxy in front of the gathering service and use unleash there to split the traffic among the canary and live versions. IMO, too much effort for the same benefits.
OK, understood. Thanks.

File: docs/canary_releases.md (hunk 6 above)
Comment (juandspy) [reply docs/canary_releases.md:70]: There are 2 options: 
1. Fetching the rules versions on start up: it's the easiest way of implementing feature flags for this case. You are right, it's riskier.
2. Fetching the rules on build: basically setting the canary and stable branches in the Dockerfile.
It's on us to decide. The advantage of the first one is that we don't need any service rebuild -> faster deployments

File: docs/canary_releases.md (hunk 1 above)
Comment (juandspy) [reply docs/canary_releases.md:60]: I wouldn't add it and keep this one as the "standard" and recommended path. It's the safest IMO.

**Comment by juandspy** [comment]
@ikerreyes 
> Because we might not control everything everyone is doing. If the process gets automated one day (hopefully), it can happen. Or even if manual, people doing the trigger and approval can be different than those who did for the 1st and the 2nd, or just by mistake. I want to know what will happen and what we can do to prevent, if needed.

I think parallel canaries doesn't make much sense here as the stable and canary versions are defined via some variables. If someone releases a new version, these variables will be replaced and the sequence will start from the beginning.
Also, when we were discussing with the rules team, they mentioned the frequency of new conditions will be the same as new rules i.e. bi-weekly

File: docs/canary_releases.md (hunk 1 above)
Comment (ikerreyes) [reply docs/canary_releases.md:60]: It is fine, I just want to have documented that if we eventually remove the canary version without changing the population, everything should be working and there is no an outage. It can be a footnote or a sidenote or something else in some other doc. But just to have it as reference.

**Comment by ikerreyes** [comment]
> @ikerreyes
> 
> > Because we might not control everything everyone is doing. If the process gets automated one day (hopefully), it can happen. Or even if manual, people doing the trigger and approval can be different than those who did for the 1st and the 2nd, or just by mistake. I want to know what will happen and what we can do to prevent, if needed.
> 
> I think parallel canaries doesn't make much sense here as the stable and canary versions are defined via some variables. If someone releases a new version, these variables will be replaced and the sequence will start from the beginning. Also, when we were discussing with the rules team, they mentioned the frequency of new conditions will be the same as new rules i.e. bi-weekly

They are not parallel but rather overlapping. Frequency is fine till there are emergency releases and other unexpected conditions. I just want to have that covered. If it is just a replacement of variables and things will start from the beginning, then I think is fine.

**Comment by juandspy** [comment]
> > @ikerreyes
> > > Because we might not control everything everyone is doing. If the process gets automated one day (hopefully), it can happen. Or even if manual, people doing the trigger and approval can be different than those who did for the 1st and the 2nd, or just by mistake. I want to know what will happen and what we can do to prevent, if needed.
> > 
> > 
> > I think parallel canaries doesn't make much sense here as the stable and canary versions are defined via some variables. If someone releases a new version, these variables will be replaced and the sequence will start from the beginning. Also, when we were discussing with the rules team, they mentioned the frequency of new conditions will be the same as new rules i.e. bi-weekly
> 
> They are not parallel but rather overlapping. Frequency is fine till there are emergency releases and other unexpected conditions. I just want to have that covered. If it is just a replacement of variables and things will start from the beginning, then I think is fine.

I added a text explaining overlapping canaries. TY

File: docs/canary_releases.md (hunk 4 above)
Comment (jholecek-rh) [reply docs/canary_releases.md:38]: That wasn't the point. The point was that if you roll out the canary without checking the Unleash configuration, the canary version would be served immediately to an unknown percentage of clusters (whatever percentage was left there from the previous deployment). We don't have to reset the percentage to 0, but checking what the percentage is should be a documented step in the process.

File: docs/canary_releases.md (hunk 3 above)
Comment (jholecek-rh) [reply docs/canary_releases.md:49]: I think Iker's point was that it should be obvious from the diagram what the behavior is. [These changes](https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/pull/234#discussion_r1675783452) should help with that.

File: docs/canary_releases.md (hunk 7 above)
Comment (jholecek-rh) [reply docs/canary_releases.md:55]: That's OK. RT and PT are two separate actors. The diagram should capture whenever either one needs to take an action.

File: docs/canary_releases.md (hunk 3 above)
Comment (ikerreyes) [reply docs/canary_releases.md:49]: Yes, thanks.

File: docs/canary_releases.md (hunk 6 above)
Comment (juandspy) [reply docs/canary_releases.md:73]: > Anything within 20-30 minutes should be fine

I'd say even less. Let's go with the less risk alternative.

**Comment by ikerreyes** [comment]
@juandspy Going back to this now that I have it more clear

> > Another question, what will be the cost and risk of having unleash set to X% all the time and we have always 2 deployments and we just change the versions of those?
> 
> I don't get this scenario: "we have always 2 deployments". Unleash is designed to be inside the code, so I'm not sure how to split traffic among 2 deployments. We would need to put a proxy in front of the gathering service and use unleash there to split the traffic among the canary and live versions. IMO, too much effort for the same benefits.

The idea here is to have canary population set to X% all the time, and the service exposing to all accounts stable version, if we are not in a canary or stable or canary depending on unleash. Basically doing the same but just without modifying unleash, so nobody needs to modify that part.

**Comment by juandspy** [comment]
> The idea here is to have canary population set to X% all the time, and the service exposing to all accounts stable version, if we are not in a canary or stable or canary depending on unleash. Basically doing the same but just without modifying unleash, so nobody needs to modify that part.

Yes. As we will remove the CANARY_VERSION variable, there is no risk of leaving Unleash at X%. I set it in the sequence diagram because it's like an extra step to be sure we are not exposing anything we don't want to, but we can ignore it @ikerreyes

**Review by joselsegura — APPROVED** [review]

File: docs/canary_releases.md (hunk 2 above)
Comment (juandspy) [reply docs/canary_releases.md]: updated

**Review by Bee-lee — APPROVED** [review]

**Review by matysek — APPROVED** [review]

**Review by ikerreyes — APPROVED** [review]
